automaton/src/main/java/org/killbill/automaton/DefaultStateMachineConfig.java

```diff
@@ -99,15 +96,15 @@ public StateMachine getStateMachineForState(final String stateName) throws Missi
     }
 
     public DefaultStateMachine[] getStateMachines() {
-        return stateMachines;
+        return stateMachines == null ? null : stateMachines.clone();
```

Comment:
This makes a shallow copy though - is this what we want?

Reply:
@pierre yeah you're correct, (especially in automaton related classes) I put that mainly to eliminate spotbugs errors. I'm thinking about something else (copy constructor, etc) but I'm not sure how far we should go:

- Should copy constructor sufficient?
- Since most of implementation classes contains `Externalizable`, should we [concern about security](https://christian-schneider.net/JavaDeserializationSecurityFAQ.html) issue as well?
- If we do concern about security, which approach should we use? I read about [serial killer](https://github.com/ikkisoft/SerialKiller) for blacklist/whitelisting `ObjectInputStream`. But should we go with another options? (and if we should, what options do you have in mind?)
- Or should we remove `clone()` because it resource waste without any advantages anyway? And we should also ignore security issue because it handle somewhere else? (I check it used in `killbill/payment`, but I'm not checking thoroughly each classes to make sure that we handle security issue)

WDYT?

Reply:
It's mostly used in payment module (internal use-case). I'm thinking we just ignore the spotbug warning, thats could be a lot of work for not many obvious benefits...